tests/test-cluster.py

```diff
@@ -463,3 +465,139 @@ def test_no_cert_reissue_in_nodes(self):
         for vm in self.VM:
             lock_files = vm.run("ls /var/snap/microk8s/current/var/lock/")
             assert "no-cert-reissue" in lock_files.decode()
+
+
+class TestUpgradeCluster(object):
+    @pytest.fixture(autouse=True, scope="module")
+    def setup_old_versioned_cluster(self):
+        """
+        Provision VMs of the previous version and form the cluster.
+        """
+        try:
+            print("Setting up cluster of an older version")
+            if channel_to_test.startswith("latest") or "/" not in channel_to_test:
+                attempt = 0
+                release_url = "https://dl.k8s.io/release/stable.txt"
+                while attempt < 10:
+                    r = requests.get(release_url)
+                    if r.status_code == 200:
+                        last_stable_str = r.content.decode().strip()
+                        last_stable_str = last_stable_str.replace("v", "")
+                        last_stable_str = ".".join(last_stable_str.split(".")[:2])
+                        break
+                    else:
+                        time.sleep(3)
+                        attempt += 1
+
+            track, _ = channel_to_test.split("/")
+            if track == "latest":
+                track = last_stable_str
+
+            # For eksd and stable tracks, we need a previous version on these tracks.
+            # Eg, to test 1.24-eksd, we need 1.23-eksd track.
+            if "-" in track:
+                track = str(float(track.split("-")[0]) - 0.01) + "-" + track.split("-")[1]
+            else:
+                track = float(track) - 0.01
+            older_version = str(track) + "/" + "stable"
+            print("Old version is {}".format(older_version))
+
+            type(self).VM = []
+            if not reuse_vms:
+                size = 1
+                for i in range(0, size):
+                    print("Creating machine {}".format(i))
+                    vm = VM(backend)
+                    vm.setup(older_version)
+                    print("Waiting for machine {}".format(i))
+                    vm.run("/snap/bin/microk8s.status --wait-ready --timeout 120")
+                    self.VM.append(vm)
+            else:
+                for vm_name in reuse_vms:
+                    vm = VM(backend, vm_name)
+                    vm.setup(older_version)
+                    self.VM.append(vm)
+
+            # Form cluster
```

Comment:
If we have a cluster of size one we do not need to join nodes. The cluster is already formed.